server/pbenchinacan/run-pbench-in-a-can

```diff
@@ -24,19 +24,41 @@ export PB_SERVER_IMAGE_PULL_POLICY="${PB_SERVER_IMAGE_PULL_POLICY:-${PB_COMMON_I
 # Directory to use for the fully built dashboard code.
 export PB_DASHBOARD_DIR="${PB_DASHBOARD_DIR:-${PWD}/dashboard/build/}"
 
+# Keycloak realm and client IDs to be used by the load_keycloak.sh script and
+# the pbench-server.cfg file.
 export KEYCLOAK_REALM=${KEYCLOAK_REALM:-"pbench-server"}
 export KEYCLOAK_CLIENT=${KEYCLOAK_CLIENT:-"pbench-client"}
 
-# Note:  the value of PB_HOST_IP will be used to generate the TLS certificate
-# and so it (not `localhost`) must also be used to access the Pbench Server;
-# otherwise, the TLS validation will fail due to a host mismatch.
-if [[ -z "${PB_HOST_IP}" ]]; then
-    host_ip_list=$(hostname -I)
-    PB_HOST_IP=${host_ip_list%% *}
-    export PB_HOST_IP
-fi
+# Name or IP address to be used by the client to access the Pbench Server, to
+# load the Dashboard, and to interface with the canned Keycloak server
+host_name=${PB_HOST:-localhost}
 
-host_name=${PB_HOST_NAME:-$(hostname --fqdn)}
+# Set a value for the -addext "subjectAltName=..." option to the cert creation.
+# Set the initial value "manually" and skip the "dummy" value to make the
+# comma-separated concatenations work nicely.
+#
+# Notes about the use of `readarray`:  in order for the output array to be
+# visible to the rest of the commands, the `readarray` command must not be
+# executed in a subprocess, and therefore it cannot be on the receiving end
+# of a pipe.  Instead we execute the `hostname` command in a "command
+# expansion" inside a "here string".  Unfortunately, this causes a newline
+# to be appended to the output, so that if we divide the result on spaces
+# we end up with an extra array entry containing just a newline.  So, instead
+# we translate the spaces into newlines ("squeezing" out any repeated
+# delimiters) and use the `readarray` default delimiter which is the newline.
+subj_alt_name="DNS.1:localhost"
+readarray -t cert_hostnames <<< "$(echo "dummy $(hostname -A)" | tr -s ' ' '\n')"
+for ((i=1; i < ${#cert_hostnames[*]}; i++)); do
+  subj_alt_name+=", DNS.$((i+1)):${cert_hostnames[i]}"
+done
```

Comment:
If you really wanted a "+1" here, you could supplement this with the IPv6 localhost, `::`. (And also `https://[::]:8443` on the redirects, I suppose... but now I'm just throwing chum...)

Reply:
Aside from the fact that `::` is an _address_, not a hostname (so it would be handled by the code _after_ the loop you cite above), I think that `::` is the IPv6 equivalent of the IPv4 `0.0.0.0`.  If you are trying to hit `localhost`, then I think you need to use `::1`.

I don't know whether we need to list that explicitly in the cert (although it probably won't hurt), but I expect we'll need it in the Keycloak redirects list.

Reply:
> Aside from the fact that `::` is an _address_, not a hostname (so it would be handled by the code _after_ the loop you cite above), I think that `::` is the IPv6 equivalent of the IPv4 `0.0.0.0`. If you are trying to hit `localhost`, then I think you need to use `::1`.
> 
> I don't know whether we need to list that explicitly in the cert (although it probably won't hurt), but I expect we'll need it in the Keycloak redirects list.

Ay, you're correct there. But the cert failure shows the list of allowed names, and it's translated `localhost` to `127.0.0.1` so, yeah, we'd also need `::1`. This wouldn't be an issue, I think, except for the npm dev-mode if we direct `PBENCH_SERVER=https://\[::1\]:8443 npm run dev`, which will currently complain.

But, yeah, to load `https://[::1]:8443/dashboard/` directly in the browser, even without the express reflector involved, we'll need the redirect defined.

Again, not a really big deal now, but if we ever went IPv6-only, it might be.

Reply:
> it's translated localhost to `127.0.0.1` so, yeah, we'd also need `::1`

I think the way that `localhost` is translated is controlled by `/etc/hosts`, so...if you're willing to muck with that, I think you can get the translation to go the other way.  😇 

In terms of the cert, I explicitly included `127.0.0.1` in the list; I'm now in the process of trying to figure out how to include `::1`, but the colons are presenting a challenge again.